### PR TITLE
MGDOBR-199 MC: Avoid keeping the DB transaction open when calling the create connectors API

### PR DESCRIPTION
**Pull request by lucamolteni**
[MGDOBR-199](https://issues.redhat.com/browse/MGDOBR-199) 

This PR is just to avoid keeping the DB transaction open and doesn't offer a full reconcile loop for stale resources

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config)
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

### CODE REVIEW

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsOrchestrateImpl.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    @Inject
+    ConnectorsDAO connectorsDAO;
+
+    @Inject
+    EventBus bus;
+
+    @Override
+    @Transactional(Transactional.TxType.NEVER)
+    public Uni<List<Message<ConnectorEntity>>> updatePendingConnectors() {
+        // assume this class is not a singleton as we could have multiple pods running this
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsOrchestrateImpl.java:31]: @lucamolteni I assume this is where the conversation around "locking" came in e.g we have two pods, both pick up the need to deploy the connector at the same time and therefore we need to ensure only 1 does the work.

The alternative here is that we use a Kafka Topic rather than our DB. So the need to deploy a Connector is an Event that we place onto a Kafka Topic. All our consumers of that Topic have the same consumer group id so only one of them ever consumes the Event. That would mean that we could have many pods processing work to deploy connectors, and only one of them ever doing the actual DB update when things have completed.

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+public abstract class AbstractConnectorWorker<R> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractConnectorWorker.class);
+
+    @Inject
+    EntityManager entityManager;
+
+    // Could be the POD name
+    String workerId = UUID.randomUUID().toString();
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java:25]: @lucamolteni Not sure this is going to work. I suspect it will just mean that we end up with orphaned items in the work queue. Imagine the following workflow:

- workerId that is random selects claims item
- workerId does first part of provisioning of connector
- workerId crashes

A new workerId is spawned by K8S. What happens to the item in the queue that was claimed by the previous worker? 

Using the Pod name as the `workerId` doesn't help as they are generated randomly via K8S controller when it creates them for the `Deployment` of the Fleet Manager.

File: infra/src/main/java/com/redhat/service/bridge/infra/models/dto/ConnectorStatus.java [hunk 3]
```diff
@@ -4,9 +4,9 @@
 // https://issues.redhat.com/browse/MGDOBR-195
 public enum ConnectorStatus {
     REQUESTED,
```
Comment (r00ta) [inline infra/src/main/java/com/redhat/service/bridge/infra/models/dto/ConnectorStatus.java:6]: You might want to align to the new statuses (from MK) https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/333

File: manager/src/main/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPI.java [hunk 4]
```diff
@@ -81,7 +85,9 @@ public Response getProcessors() {
         String shardId = identityResolver.resolve(jwt);
         failIfNotAuthorized(shardId);
         LOGGER.info("Request from Shard for Processors to deploy or delete.");
-        return Response.ok(processorService.getProcessorByStatusesAndShardId(statuses, shardId)
+        List<Processor> processorToDeployOrDelete = processorService.getProcessorByStatusesAndShardId(statuses, shardId);
+        LOGGER.info("Found {} processor(s) to deploy or delete", processorToDeployOrDelete.size());
```
Comment (r00ta) [inline manager/src/main/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPI.java:89]: If you log this for Processor you might add it also for the Bridge ingresses

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/creation/CreateConnectorWorker.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    @ConsumeEvent(value = KAFKA_TOPIC_CREATED_EVENT, blocking = true)
+    public void consume(ConnectorEntity connectorEntity) {
+        execute(connectorEntity);
+    }
+
+    @Override
+    protected Connector callExternalService(ConnectorEntity connectorEntity) {
+        JsonNode payload = connectorEntity.getDefinition();
+        String newConnectorName = connectorEntity.getName();
```
Comment (r00ta) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/creation/CreateConnectorWorker.java:53]: Sorry for the stylish nitpick (feel free to ignore it): you define those variables here but you use them only once and after few lines. Maybe it would be more readable to do 
```java
createConnectorRequest.setName(connectorEntity.getName());
```

instead of 

```java
String newConnectorName = connectorEntity.getName();
String connectorType = connectorEntity.getConnectorType();
ConnectorRequest createConnectorRequest = new ConnectorRequest();
createConnectorRequest.setName(newConnectorName);
```

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/creation/CreateConnectorWorker.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    @ConfigProperty(name = "managed-connectors.kafka.client.id")
+    String serviceAccountId;
+
+    @ConfigProperty(name = "managed-connectors.kafka.client.secret")
+    String serviceAccountSecret;
+
+    @Inject
+    ConnectorsApiClient connectorsApiClient;
+
+    @ConsumeEvent(value = KAFKA_TOPIC_CREATED_EVENT, blocking = true)
```
Comment (r00ta) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/creation/CreateConnectorWorker.java:45]: In the other workers you use `Events.CONSTANT` and here you use static import (also in other workers), maybe one single approach would improve readability

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsOrchestrate.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package com.redhat.service.bridge.manager.connectors;
+
+import java.util.List;
+
+import com.redhat.service.bridge.manager.models.ConnectorEntity;
+
+import io.smallrye.mutiny.Uni;
+import io.vertx.mutiny.core.eventbus.Message;
+
+public interface ConnectorsOrchestrate {
```
Comment (r00ta) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsOrchestrate.java:10]: `ConnectorsOrchestrator`?

File: manager/src/main/java/com/redhat/service/bridge/manager/dao/ConnectorsDAO.java [hunk 8]
```diff
@@ -20,9 +20,13 @@ public ConnectorEntity findByProcessorIdAndName(String processorId, String name)
         return singleResultFromList(find("#CONNECTORENTITY.findByProcessorIdAndName", p));
     }
 
-    public ConnectorEntity findByProcessorId(String processorId) {
+    public List<ConnectorEntity> findUnprocessed() {
+        return find("#CONNECTORENTITY.findUnprocessed").list();
+    }
+
+    public List<ConnectorEntity> findByProcessorId(String processorId) {
```
Comment (r00ta) [inline manager/src/main/java/com/redhat/service/bridge/manager/dao/ConnectorsDAO.java:27]: Is this going to return always a list with just one item?

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+                deleteEntity(connectorEntity);
+            }
+        } catch (Exception e) {
+
+            errorWhileCalling(e, connectorEntity);
+        }
+    }
+
+    @Transactional(Transactional.TxType.REQUIRES_NEW)
+    public ConnectorEntity claimConnector(ConnectorEntity currentEntity) {
```
Comment (r00ta) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java:80]: I think there is a concurrency here. Think about the following scenario: 
1) Pod A runs `ConnectorsOrchestrator.updatePendingConnectors` and with `ConnectorDAO.findUnprocessed` finds the connector X. It sends to its internal bus the request. 
2) Pod B does the same and finds the connector X. The request is sent to its internal bus as well. 
3) Pod A starts processing and calls this `claimConnector` which sets `mergedCE.setWorkerId(workerId);`
4) Pod B does the same and overrides with its `mergedCE.setWorkerId(workerId);`

In short, I think this `claimConnector` is not really preventing multiple pods to take ownership of the resource since it is not checking that the `workerId` is still `null`.  Or am I missing something?

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsOrchestrateImpl.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Inject
+    ConnectorsDAO connectorsDAO;
+
+    @Inject
+    EventBus eventBus;
+
+    @Override
+    @Transactional(Transactional.TxType.NEVER)
+    public Uni<List<Message<ConnectorEntity>>> updatePendingConnectors() {
```
Comment (r00ta) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsOrchestrateImpl.java:37]: I think a pending connector might be also a connector that has the `workerId` set but for any reason that worker did not release the lock (it crashed, it was restarted, any reason). Currently `findUprocessed` is not returning such processors

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java [hunk 11]
```diff
@@ -112,13 +120,19 @@ public Processor createProcessor(String bridgeId, String customerId, ProcessorRe
         ProcessorDefinition definition = new ProcessorDefinition(requestedFilters, requestedTransformationTemplate, requestedAction, resolvedAction);
         newProcessor.setDefinition(definitionToJsonNode(definition));
 
-        processorDAO.persist(newProcessor);
+        Optional<ConnectorEntity> persist = persistProcessorCreateConnector(newProcessor, actionProvider, resolvedAction);
 
-        connectorService.createConnectorIfNeeded(resolvedAction, newProcessor, actionProvider);
+        persist.ifPresent(c -> eventBus.requestAndForget(CONNECTOR_CREATED_EVENT, c));
```
Comment (r00ta) [inline manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:125]: I'm wondering if it's responsability of the `connectorService` to fire the event

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java [hunk 12]
```diff
@@ -160,19 +174,24 @@ public Long getProcessorsCount(String bridgeId, String customerId) {
...
-        Processor processor = processorDAO.findByIdBridgeIdAndCustomerId(processorId, bridgeId, customerId);
-
-        ProcessorDefinition processorDefinition = jsonNodeToDefinition(processor.getDefinition());
-        ActionProvider actionProvider = actionProviderFactory.getActionProvider(processorDefinition.getRequestedAction().getType());
-
-        connectorService.deleteConnectorIfNeeded(processorDefinition.getResolvedAction(), processor, actionProvider);
+        List<ConnectorEntity> connectorEntitiesToBeDeleted = updateProcessorConnectorForDeletionOnDB(bridgeId, processorId, customerId);
+        connectorEntitiesToBeDeleted.forEach(c -> {
+            LOGGER.info("Firing deletion event for entity: " + c);
+            eventBus.requestAndForget(Events.CONNECTOR_DELETED_EVENT, c);
```
Comment (r00ta) [inline manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:182]: same here

File: manager/src/main/java/com/redhat/service/bridge/manager/models/Processor.java [hunk 13]
```diff
@@ -29,7 +32,14 @@
         @NamedQuery(name = "PROCESSOR.findByBridgeIdAndName",
                 query = "from Processor p where p.name=:name and p.bridge.id=:bridgeId"),
         @NamedQuery(name = "PROCESSOR.findByStatusesAndShardId",
```
Comment (r00ta) [inline manager/src/main/java/com/redhat/service/bridge/manager/models/Processor.java:34]: I'd say now the signature is more something like `findByStatusesAndShardIdWithReadyDependencies`

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java [hunk 14]
```diff
@@ -112,13 +120,19 @@ public Processor createProcessor(String bridgeId, String customerId, ProcessorRe
...
-        processorDAO.persist(newProcessor);
+        Optional<ConnectorEntity> persist = persistProcessorCreateConnector(newProcessor, actionProvider, resolvedAction);
 
-        connectorService.createConnectorIfNeeded(resolvedAction, newProcessor, actionProvider);
+        persist.ifPresent(c -> eventBus.requestAndForget(CONNECTOR_CREATED_EVENT, c));
 
         return newProcessor;
     }
 
+    @Transactional
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:130]: I don't think the `@Transactional` annotation works on private methods. Is there any reason why you removed it from the public `createProcessor` method?

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java [hunk 15]
```diff
@@ -160,19 +174,24 @@ public Long getProcessorsCount(String bridgeId, String customerId) {
...
-        connectorService.deleteConnectorIfNeeded(processorDefinition.getResolvedAction(), processor, actionProvider);
+        List<ConnectorEntity> connectorEntitiesToBeDeleted = updateProcessorConnectorForDeletionOnDB(bridgeId, processorId, customerId);
+        connectorEntitiesToBeDeleted.forEach(c -> {
+            LOGGER.info("Firing deletion event for entity: " + c);
+            eventBus.requestAndForget(Events.CONNECTOR_DELETED_EVENT, c);
+        });
+    }
 
+    @Transactional
+    public List<ConnectorEntity> updateProcessorConnectorForDeletionOnDB(String bridgeId, String processorId, String customerId) {
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:187]: Is there any reason why this needs to be a separate method to `deleteProcessor`?

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java (hunk 12 above)
Comment (robpblake) [reply manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:182]: I agree with @r00ta comment.

 I feel it would be cleaner for our first implementation to have a `createConnector()` and `deleteConnector` methods on the `ConnectorService` to align with the `ProcessorService` and `BridgeService` interfaces. 
 
 Within the `createConnector` method, if we don't need to create a Connector, it's essentially a null-op. Otherwise we do what's required and fire the event from within the `createConnector` method.

File: manager/src/main/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPI.java (hunk 4 above)
Comment (robpblake) [reply manager/src/main/java/com/redhat/service/bridge/manager/api/internal/ShardBridgesSyncAPI.java:89]: If we decide to keep this can we also include the `shardId` in the log statement. At the minute it doesn't tell us for which Shard e.g:

`Found '4' processor(s) to deploy or delete for shard 'foo'`

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+        } catch (Exception e) {
+            // Cannot claim, return the unmodified entity
+            LOGGER.trace("Worker {} cannot claim Entity: {} with error: {} ", workerId, workerName(), e);
+            return;
+        }
+
+        try {
+            R serviceResponse = callExternalService(connectorEntity);
+
+            if (!shouldDeleteAfterSuccess()) {
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java:66]: Can we reverse the logic here to make it easier to follow e.g.:

```
if(shouldDeleteAfterSuccess()) {
  deleteEntity();
} else {
...
}

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java (hunk 9 above)
Comment (robpblake) [reply manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java:80]: We have a `@Version` annotation on the `ConnectorEntity` to enforce optimistic locking. With that only one transaction will be able to successfully commit the claim on the entity. The other will fail because the other transaction committed first and updated the version of the Entity. 

Essentially the `@Version` annotation makes the logic of the db write something like:

```
int updatedRows = persistEntity(entity);
if(updatedRows == 0){
throw OptimisticLockingException()
}
```

The SQL generated for the update looks something like `update table connector set id=1 where version=1`

If the `version` has changed then the number of rows updated is 0 and the commit fails.

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsService.java [hunk 17]
```diff
@@ -9,11 +10,9 @@
 
 public interface ConnectorsService {
 
-    Optional<ConnectorEntity> createConnectorIfNeeded(BaseAction resolvedAction,
+    Optional<ConnectorEntity> persistConnectorIfNeeded(BaseAction resolvedAction,
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsService.java:13]: @lucamolteni Can we call this `createConnector` to align us with our other Service interfaces. The fact that it is returning an `Optional` to me means we don't need the `IfNeeded` suffix.

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsService.java [hunk 18]
```diff
@@ -9,11 +10,9 @@
...
             Processor processor,
             ActionProvider actionProvider);
 
-    void deleteConnectorIfNeeded(BaseAction resolvedAction,
-            Processor processor,
-            ActionProvider actionProvider);
+    List<ConnectorEntity> deleteConnectorIfNeeded(Processor processor);
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsService.java:17]: @lucamolteni How comes a create returns a single `ConnectorEntity`, but a delete returns a list of `ConnectorEntity` ? Is there a scenario where the there is a one-to-many relationship between `Processor` and `Connector`?

File: manager/src/main/java/com/redhat/service/bridge/manager/models/Processor.java [hunk 19]
```diff
@@ -29,7 +32,14 @@
...
-                query = "from Processor p join fetch p.bridge where p.status in (:statuses) and p.bridge.status='AVAILABLE' and p.shardId=:shardId"),
+                query = "select p " +
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/bridge/manager/models/Processor.java:35]: @lucamolteni I just want to check my memory of our discussions: We said that we would make this more generic in the future so that we can more easily determine if all resources for either a BridgeInstance or a Processor are ready. This is the first iteration of this.

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+    // Used for workers that should delete the entity after successfully calling the service.
+    // By default, after the service call updateEntityForSuccess is called, with this physical deletion will occur
+    protected boolean shouldDeleteAfterSuccess() {
+        return false;
+    }
+
+    // Could be the k8s pod name
+    String workerId = UUID.randomUUID().toString();
+
+    @Transactional(Transactional.TxType.NEVER)
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java:50]: @lucamolteni I'd suggest adding a comment here as to why `TxType.NEVER` is important on this method for other people coming to the code who are perhaps less experienced with JPA/Hibernate than you.

**Review by robpblake — COMMENTED** [review]
Thanks @lucamolteni . Some feedback for you.

This is definitely moving us in the right direction, so thank you for getting it to this point 👍 

cheers,

Rob

File: manager/src/main/java/com/redhat/service/bridge/manager/dao/ConnectorsDAO.java (hunk 8 above)
Comment (lucamolteni) [reply manager/src/main/java/com/redhat/service/bridge/manager/dao/ConnectorsDAO.java:27]: It's a `OneToOne` mapped with a `OneToMany/ManyToOne` so theoretically yes, even though in practice it's impossible. The reason for which I haven't used `OneToOne` is that in Hibernate they're usually complicated and it should be avoided (please excuse me why I don't write the exact reasons , I need to check again but it's something I've memorized) :)

I'd rather not assume in the code that we might not have by error two connectors per processor, so I think it's better to map it with a list anyway.

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java (hunk 9 above)
Comment (lucamolteni) [reply manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java:80]: `CONNECTORENTITY.findUnprocessed` checks for the unprocessed entities (those who don't have the workerId) and the claim is a lock that prevents two worker from working on the same entity

Comment (lucamolteni) [reply manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java:80]: Closing this as it should be cleared

File: manager/src/main/java/com/redhat/service/bridge/manager/models/Processor.java (hunk 19 above)
Comment (lucamolteni) [reply manager/src/main/java/com/redhat/service/bridge/manager/models/Processor.java:35]: As this method is getting more and more specific, I think we should consider it in the future. I'm closing it for now

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsOrchestrateImpl.java (hunk 10 above)
Comment (lucamolteni) [reply manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsOrchestrateImpl.java:37]: Yes this will be adressed in https://issues.redhat.com/browse/MGDOBR-155

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java (hunk 14 above)
Comment (lucamolteni) [reply manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:130]: Good catch, it was the ConnectorService opening a transaction because we changed from `MANDATORY` to `REQUIRES_NEW` in the previous PR. I'm setting Connector's back to `MANDATORY` with a comment

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java (hunk 15 above)
Comment (lucamolteni) [reply manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:187]: We don't want to fire the event on the service bus while being in a transaction, so the method that writes on the DB is separated with its own `@Transactional` annotation, when it returns the event is fired

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java (hunk 11 above)
Comment (lucamolteni) [reply manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:125]: Doing do will need to call yet another method on that service. A separation of service mapping 1:1 the name with the entities won't scale when we modify one than more entity.

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java (hunk 12 above)
Comment (lucamolteni) [reply manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:182]: I think we should revisit the separation of Services - I don't think that having a separate service per entity is useful at all, especially in features dealing with multiple entity types.

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsService.java (hunk 18 above)
Comment (lucamolteni) [reply manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsService.java:17]: It's a `OneToOne` mapped as a `OneToMany/ManyToOne`, we cannot exclude that for a bug multiple connectors where created, so it's safer to delete them all anyway

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsServiceImpl.java [hunk 21]
```diff
@@ -70,91 +44,38 @@
...
-        // https://issues.redhat.com/browse/MGDOBR-198
-        // this is currently ignored in the Connectors API
-        kafka.setId(KAFKA_ID_IGNORED);
+        List<ConnectorEntity> optionalConnector = connectorsDAO.findByProcessorId(processor.getId());
 
-        createConnectorRequest.setKafka(kafka);
+        return optionalConnector.stream().peek(c -> {
+            String connectorId = c.getId();
+            c.setDesiredStatus(ConnectorStatus.DELETED);
+            LOGGER.info("connector with id '{}' has been market for deletion", connectorId);
```
Comment (jomarko) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/ConnectorsServiceImpl.java:74]: small typo here

File: manager/src/main/java/com/redhat/service/bridge/manager/models/Processor.java [hunk 22]
```diff
@@ -28,8 +31,15 @@
...
-                query = "from Processor p join fetch p.bridge where p.status in (:statuses) and p.bridge.status='AVAILABLE' and p.shardId=:shardId"),
+        @NamedQuery(name = "PROCESSOR.findByStatusesAndShardIdWithReadyDependencies",
+                query = "select p " +
+                        "from Processor p " +
+                        "join fetch p.bridge " +
+                        "left join p.connectorEntities as c " +
+                        "where p.status in (:statuses) " +
+                        "and p.bridge.status='AVAILABLE' " +
+                        "and p.shardId=:shardId " +
+                        "and (c is null or c.status = 'READY')"),
```
Comment (jomarko) [inline manager/src/main/java/com/redhat/service/bridge/manager/models/Processor.java:42]: Do not we have some enums for AVAILABLE and READY? just asking as saw some PR where some enums were renamed. So maybe using enums here could help if some status is renamed in future?

**Review by jomarko — COMMENTED** [review]
thank you for a PR two newbie comments from me, probably not blocking from merging

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    // Used for workers that should delete the entity after successfully calling the service.
+    // By default, after the service call updateEntityForSuccess is called, with this physical deletion will occur
+    protected boolean shouldDeleteAfterSuccess() {
+        return false;
+    }
+
+    // Could be the k8s pod name
+    String workerId = UUID.randomUUID().toString();
```
Comment (danielezonca) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java:48]: Do you have a followup ticket to address/review this?

**Review by danielezonca — APPROVED** [review]
LGTM

File: manager/src/main/java/com/redhat/service/bridge/manager/models/Processor.java (hunk 22 above)
Comment (lucamolteni) [reply manager/src/main/java/com/redhat/service/bridge/manager/models/Processor.java:42]: We have a generic `BridgeStatus` enum for the entities having the same lifecycle

File: infra/src/main/java/com/redhat/service/bridge/infra/models/dto/ConnectorStatus.java [hunk 24]
```diff
@@ -3,10 +3,16 @@
 // TODO-MC review statuses
 // https://issues.redhat.com/browse/MGDOBR-195
 public enum ConnectorStatus {
-    REQUESTED,
-    PROVISIONING,
-    AVAILABLE,
-    DELETION_REQUESTED,
-    DELETED,
-    FAILED
+    ACCEPTED("accepted"),
```
Comment (r00ta) [inline infra/src/main/java/com/redhat/service/bridge/infra/models/dto/ConnectorStatus.java:6]: Unfortunately you have to migrate the existing rows in the database otherwise the demo cluster is broken. 
In short you have to do something like https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/fb7d556e761207bd649b1c20ba5d50a5f16c1226/manager/src/main/resources/db/migration/0.1/V0.1.6__migrate_statuses.sql

**Review by r00ta — APPROVED** [review]
LGTM thanks for the clarifications

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+        return mergedCE;
+    }
+
+    @Transactional(Transactional.TxType.REQUIRES_NEW)
+    public ConnectorEntity successfullyCalledService(ConnectorEntity connectorEntity, R serviceResponse) {
+        ConnectorEntity updatedEntity = updateEntityForSuccess(connectorEntity, serviceResponse);
+        connectorEntity.setWorkerId(null);
+        connectorEntity.setModifiedAt(ZonedDateTime.now(ZoneOffset.UTC));
+
+        LOGGER.info("Worker {} successfully called service with response : {} on entity: {} ", workerName(), serviceResponse, updatedEntity);
```
Comment (sutaakar) [inline manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java:97]: Is it safe to log complete response and entity? It may have sensitive informations.

**Review by sutaakar — APPROVED** [review]
Looks good to me, though I wasn't able to test it properly due to MK issues.

File: manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java (hunk 25 above)
Comment (lucamolteni) [reply manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java:97]: I don't think it's much more sensitive than the actual OpenBridge data such as Processor ID, so I'll leave it now as it's useful for debugging
